text/0253-cdk-metadata-v2.md

```diff
@@ -0,0 +1,267 @@
+---
+feature name: CDK Metadata v2
+start date: 2020-09-23
+rfc pr: (leave this empty)
+related issue: #253
+---
+
+# Abstract
+
+The CDK Metadata resource will start tracking individual constructs used in a
+Stack's synthesis. For every construct used in a Stack from an AWS-authored
+library we will collect the library name, version and qualified class name
+(using a generated lookup table to obtain submodule names) and submit these
+to the metadata resource in a gzipped blob.
+
+# Motivation
+
+The goal of CDK Metadata is twofold:
+
+* Track what library versions people are using, so that if there are severy (security)
+  problems with a library, we can contact them in a targeted fashion and suggest
+  they upgrade.
+* See what construct libraries are being used by people to inform future development
+  activities (example insight: are there areas where people are using L1s
+  instead of L2s, signaling the need to invest more effort into L2
+  development).
+
+The current CDK Metadata is based on libraries loaded into the NodeJS process. When
+we move to CDK v2, the entire CDK library will be vended as one library, destroying
+our ability to measure in the same way.
+
+At the same time, our current method fails in the face of multiple distinct
+stacks in the same application. For example, in the case of a CDK Pipelines
+application, the `@aws-cdk/pipelines` library would be counted `N+1` times:
+in addition to the pipeline stack itself, once for every environment deployed
+via that pipeline because the library happened to be in memory while those
+stacks were being synthesized.
```

Comment:
Not sure that I agree. Let's say a CDK Pipelines app contains a `PipelineStack` and an `ApiGateway` stack (the latter to be deployed using the former).

Would you prefer this be counted as:

* 1 time a stack using `@aws-cdk/pipelines` (the `PipelineStack`)
* 1 time a stack using `@aws-cdk/api-gateway` (the `ApiGateway` stack)

Or

* 2 times a stack using `@aws-cdk/pipelines` (both stacks, because they both happen to be synthesized by an application that uses `@aws-cdk/pipelines`).
* 2 times a stack using `@aws-cdk/api-gateway` (both stacks, because they both happen to be synthesized by an application that uses `@aws-cdk/pipelines`).

You seem to be arguing you prefer those stacks be counted as the latter, but I'm not sure that would be all that intuitive to most people.